src/lpf2hub.ts

```diff
@@ -439,174 +453,182 @@ export class LPF2Hub extends Hub {
-                    this._lastTiltY = tiltY;
-                    this.emit("tilt", port.id, this._lastTiltX, this._lastTiltY, this._lastTiltZ);
-                    break;
-                }
-                case Consts.DeviceType.POWERED_UP_REMOTE_BUTTON: {
-                    switch (data[4]) {
-                        case 0x01: {
-                            this.emit("button", port.id, Consts.ButtonState.UP);
-                            break;
-                        }
-                        case 0xff: {
-                            this.emit("button", port.id, Consts.ButtonState.DOWN);
-                            break;
-                        }
-                        case 0x7f: {
-                            this.emit("button", port.id, Consts.ButtonState.STOP);
-                            break;
-                        }
-                        case 0x00: {
-                            this.emit("button", port.id, Consts.ButtonState.RELEASED);
-                            break;
-                        }
-                    }
-                    break;
-                }
-                case Consts.DeviceType.DUPLO_TRAIN_BASE_COLOR: {
-                    if (data[4] <= 10) {
-                        this.emit("color", port.id, data[4]);
-                    }
-                    break;
-                }
-                case Consts.DeviceType.DUPLO_TRAIN_BASE_SPEEDOMETER: {
-                    /**
-                     * Emits on a speed change.
-                     * @event LPF2Hub#speed
-                     * @param {string} port
-                     * @param {number} speed
-                     */
-                    const speed = data.readInt16LE(4);
-                    this.emit("speed", port.id, speed);
-                    break;
+        if (port && port.connected && port.mode && port.modes[port.mode]) {
+            const mode = port.modes[port.mode];
+            const values = [];
+
+            for (let index = 4; index < data.length; index += Consts.VALUE_SIZE[mode.valueType]) {
+                switch (mode.valueType) {
+                    case Consts.ValueType.Int8:
+                        values.push(data.readInt8(index));
+                        break;
+                    case Consts.ValueType.Int16:
+                        values.push(data.readInt16LE(index));
+                        break;
+                    case Consts.ValueType.Int32:
+                        values.push(data.readInt32LE(index));
+                        break;
+                    case Consts.ValueType.Float:
+                        values.push(data.readFloatLE(index));
+                        break;
                 }
             }
+            this.emit("sensor", port.id, mode, values);
+
+            this._emitSensorEvent(port, mode, values);
         }
 
     }
 
-
+    private _emitSensorEvent(port: Port, mode: IPortMode, values: number[]) {
+        const modeToEvent: { [key: string]: string } = {
```

Comment:
Good idea with the event name map, but I think it would be cool it the users could also subscribe to ports using these names. Otherwise you need to remember that to fire `distance` events you need to use `PROX` or `LPF2-DETECT` depending on the device used.

Reply:
There is aleady `this.emit("sensor", port.id, mode, values);` and it would make a lot of duplicated events, I don't know what approach should be adopted.